### PR TITLE
Add DSL helpers to enable metrics and reports

### LINKED ISSUE #1616: Add DSL function to more easily facilitate dumping Compose compiler metrics
This requires adding a flag to the compiler which we could handle automatically.

Details on what it all means here: https://github.com/androidx/androidx/blob/androidx-main/compose/compiler/design/compiler-metrics.md

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #1616

### CODE REVIEW

**Comment by JakeWharton** [comment]
Actually this needs some work. We need to capture the Kotin compilation task name in the directory or else main/test or multiplatform will overwrite each other.

**Comment by JakeWharton** [comment]
Screenshot 2023-10-23 at 8 20 52 PM

File: redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodComposeExtension.kt [hunk 1]
```diff
@@ -39,7 +36,35 @@ public abstract class RedwoodComposeExtension
...
+   * directory. See
+   * [the compiler documentation](https://github.com/androidx/androidx/blob/androidx-main/compose/compiler/design/compiler-metrics.md#reports-breakdown)
+   * for more information about the contents.
+   *
+   * **NOTE:** This should only be enabled during investigation as it breaks the use of
+   * Gradle's build cache for this project's Kotlin compilation tasks.
+   *
+   * @see enableReports
+   */
+  public fun enableMetrics()
```
Comment (swankjesse) [inline redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodComposeExtension.kt]: Worth incorporating the scary consequence into the function name?

```
enableMetrics_ThisBreaksGradleBuildCaching_()
```

I might inadvertently check in code that calls enableMetrics(), but my code reviewer would catch that!

**Review by swankjesse — APPROVED** [review]

File: redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodComposeExtension.kt (hunk 1 above)
Comment (JakeWharton) [reply redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodComposeExtension.kt]: It wouldn't actually be the end of the world if you did so. So maybe we can find something that's still cautionary but not as in-your-face. Will sleep on it.

Comment (JakeWharton) [reply redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodComposeExtension.kt]: What about wrapping them into something else like
```groovy
redwood {
  dangerZone {
    enableMetrics()
  }
}
```
Is that enough to signal you should learn more about why it's dangerous?

Comment (swankjesse) [reply redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodComposeExtension.kt]: Ooooh I like that a lot

Comment (JakeWharton) [reply redwood-gradle-plugin/src/main/kotlin/app/cash/redwood/gradle/RedwoodComposeExtension.kt]: Done. Also added tests.

**Pull request merged**